packages/ui/src/components/Card.tsx

```diff
@@ -122,115 +109,80 @@ function useAnimatingStyles(backIsRevealed: boolean): {
   const numberOfLines = size === "regular" ? 3 : 1;
-  return (
-    promptContext && (
-      <View style={style}>
-        {promptContext.url ? (
-          <Button
-            size={size}
-            href={promptContext.url}
-            title={promptContext.title}
-            color={color}
-            numberOfLines={numberOfLines}
-            ellipsizeMode="tail"
-          />
-        ) : (
-          <Text
-            style={[
-              size === "regular"
-                ? type.label.layoutStyle
-                : type.labelTiny.layoutStyle,
-              {
-                color,
-              },
-            ]}
-            numberOfLines={numberOfLines}
-            ellipsizeMode="tail"
-          >
-            {promptContext.title}
-          </Text>
-        )}
-      </View>
-    )
-  );
+  return provenance?.title ? (
+    <View style={style}>
+      {provenance.url ? (
+        <Button
+          size={size}
+          href={provenance.url}
+          title={provenance.title}
+          color={color}
+          numberOfLines={numberOfLines}
+          ellipsizeMode="tail"
+        />
+      ) : (
+        <Text
+          style={[
+            size === "regular"
+              ? type.label.layoutStyle
+              : type.labelTiny.layoutStyle,
+            {
+              color,
+            },
+          ]}
+          numberOfLines={numberOfLines}
+          ellipsizeMode="tail"
+        >
+          {provenance.title}
+        </Text>
+      )}
+    </View>
+  ) : null;
 }
 
 function formatClozePromptContents(
-  clozeContents: string,
+  text: string,
+  component: ClozeTaskContentComponent,
   isRevealed: boolean,
-  clozeIndex: number,
 ) {
-  let matchIndex = 0;
-  let match: RegExpExecArray | null;
-
-  let output = "";
-  let previousMatchStartIndex = 0;
-  let foundSelectedClozeDeletion = false;
-  const clozeRegexp = createClozeMarkupRegexp();
-  for (; (match = clozeRegexp.exec(clozeContents)); matchIndex++) {
-    output += clozeContents.slice(previousMatchStartIndex, match.index);
-    if (matchIndex === clozeIndex) {
-      foundSelectedClozeDeletion = true;
-      if (isRevealed) {
-        // We emit the original delimeted string (i.e. "{cloze}"), which is styled in CardField.
-        output += clozeContents.slice(match.index, clozeRegexp.lastIndex);
-      } else {
-        output += clozeBlankSentinel;
-      }
+  let mutatingText: string = text;
+  for (const range of component.ranges) {
+    if (isRevealed) {
+      mutatingText =
+        mutatingText.slice(0, range.startIndex) +
+        "{" +
+        mutatingText.slice(range.startIndex, range.startIndex + range.length) +
+        "}" +
+        mutatingText.slice(range.startIndex + range.length);
     } else {
-      output += match[1]; // strip the braces
+      mutatingText =
+        mutatingText.slice(0, range.startIndex) +
+        clozeBlankSentinel +
+        mutatingText.slice(range.startIndex + range.length);
```

Comment:
Ended up not handling the hint case. I will create an issue for that when this PR lands.